rust/src/storage/file/mod.rs

```diff
@@ -101,17 +102,26 @@ impl StorageBackend for FileStorageBackend {
         if let Some(parent) = Path::new(path).parent() {
             fs::create_dir_all(parent).await?;
         }
+        let tmp_path = &format!("{}_{}", path, Uuid::new_v4().to_string());
         let mut f = fs::OpenOptions::new()
             .create(true)
             .truncate(true)
             .write(true)
-            .open(path)
+            .open(tmp_path)
             .await?;
 
         f.write_all(obj_bytes).await?;
         f.flush().await?;
-
-        Ok(())
+        drop(f);
+
+        match fs::rename(tmp_path, path).await {
```

Comment:
@zijie0 could you change this to use `rename::atomic_rename(src, dst)`?

Reply:
I just noticed our atomic rename is not async, which could block the tokio runtime and lead to latency issues in production. filed https://github.com/delta-io/delta-rs/issues/369 as a follow up for that.

Reply:
Done.

Reply:
Oh god, I made a very bad suggestion here :( I think the `fs:rename` you used here is the correct one. It does exactly what we needed, i.e. overwrite the destination if already exists. Our atomic rename should be named something else like `rename_if_not_exists` so people won't confuse it with `rename`.

Reply:
doc to the builtin rename function: https://doc.rust-lang.org/std/fs/fn.rename.html

Reply:
@houqp It seems that rename on Unix does have some atomicity issue: 

> If newpath already exists, it will be atomically replaced, so
       that there is no point at which another process attempting to
       access newpath will find it missing.  However, there will
       probably be a window in which both oldpath and newpath refer to
       the file being renamed.

But other projects like `tempfile` use it anyway.

Reply:
yes, the main goal we are trying to achieve here is to avoid reading partially written data from `newpath`/`targetpath`. It's ok for both path to exist and refer to the same file content as long as the file content itself is atomically presented to the readers.

Reply:
If @mosyp and you all agree, I may submit a new PR to use `tokio::fs::rename` in `put_obj` method.

Reply:
@zijie0 hey, I've the notification, sure, works for me!